### PR TITLE
More duration updates

### PR DESCRIPTION
**Pull request by dmlap**
Pulled in #279 so that seeking "between segments" does not cause an error. Updated duration calculation so that PTS-based results are used for the largest span of segments possible. Previously, PTS values were only used to calculate duration when they formed a continuous interval. This is a lot more conservative than necessary because PTS values are directly comparable unless there is a discontinuity between their containing segments. This method should result in more accurate calculations and get rid of some unnecessary segment iteration.

### CODE REVIEW

**Comment by dmlap** [comment]
Also, closes #314

File: src/m3u8/m3u8-parser.js [hunk 1]
```diff
@@ -375,7 +375,8 @@
 
     // the manifest is empty until the parse stream begins delivering data
     this.manifest = {
-      allowCache: true
+      allowCache: true,
+      discontinuityStarts: []
```
Comment (gkatsev) [inline src/m3u8/m3u8-parser.js]: so `discontinuityStarts` end up being a list of indexes pointing at segments that begin a discontinuity?

File: src/videojs-hls.js [hunk 2]
```diff
@@ -1148,14 +1160,15 @@ videojs.Hls.getMediaIndexByTime = function() {
...
+  var index, time = 0, segment;
 
   if (!playlist.segments || mediaIndex === 0) {
     return 0;
   }
 
   for (index = 0; index < mediaIndex; index++) {
-    time += playlist.segments[index].duration;
+    segment = playlist.segments[index];
+    time += segment.preciseDuration || segment.duration || playlist.targetDuration || 0;
```
Comment (gkatsev) [inline src/videojs-hls.js]: I thought we don't have `preciseDuration` anymore?

File: src/playlist.js [hunk 3]
```diff
@@ -5,7 +5,130 @@
...
+   * segments to examine.
+   * @param end {number} an exclusive upper boundary for the segments
+   * to examine.
+   * @param includeTrailingTime {boolean} if false, the interval between
+   * the final segment and the subsequent segment will not be included
+   * in the result
+   * @return {number} the duration between the start index and end
+   * index in seconds.
+   */
+  accumulateDuration = function(playlist, start, end, includeTrailingTime) {
```
Comment (gkatsev) [inline src/playlist.js]: so all this stuff tries to find those overlapping ranges of audio and video durations and take the greatest of them?

File: src/videojs-hls.js (hunk 2 above)
Comment (dmlap) [inline src/videojs-hls.js]: I believe this function got re-added somewhere along the way. It should be removed entirely.

File: src/playlist.js (hunk 3 above)
Comment (dmlap) [inline src/playlist.js]: Basically, yes. It tries to find the longest span of segments that have PTS info available at the start and end. It does this by:

1. Splitting the interval into ranges that do not have internal discontinuities
1. For each of these ranges, finding the earliest and latest segment with PTS info
 1. Adding up the differences between the earliest PTS and the latest PTS in each range
 1. Adding up the M3U8 based durations for any segments in the range that aren't in the interval above

File: src/m3u8/m3u8-parser.js (hunk 1 above)
Comment (dmlap) [inline src/m3u8/m3u8-parser.js]: Yes. It makes splitting intervals into continuous spans a lot simpler.

File: src/videojs-hls.js (hunk 2 above)
Comment (gkatsev) [inline src/videojs-hls.js]: looks good other than this.

**Pull request closed without merging**